language/move-prover/src/boogie_wrapper.rs

```diff
@@ -138,12 +145,107 @@ impl<'env> BoogieWrapper<'env> {
         debug!("writing boogie log to {}", boogie_log_file);
         fs::write(&boogie_log_file, &all_output)?;
 
+        // Keep a list of all functions that passed the smoke test
+        let mut smoke_test_errors = HashSet::new();
+
         for error in &errors {
-            self.add_error(error);
+            if let Some(res) = self.is_smoke_test_error(error) {
+                // Errors from smoke test checks
+                smoke_test_errors.insert(res);
+            } else {
+                // Errors from counter examples to the specification
+                self.add_error(error);
+            }
         }
+
+        // Add errors for functions with smoke tests
+        self.add_smoke_test_errors(smoke_test_errors);
+
         Ok(())
     }
 
+    /// Helper that returns `Some` tuple of `Symbol` representing the function
+    /// the error appears in and `SmokeTestKind` representing the kind of smoke test
+    /// error.
+    /// Returns None if it's not a smoke test error.
+    ///
+    /// List of smoke test errors when `pragma smoke_test = true;` is declared:
+    ///     1. Ignore specification post conditions and add `ensures $abort_flag;` to
+    ///         the function to check if the function always aborts
+    fn is_smoke_test_error(&self, error: &BoogieError) -> Option<(Symbol, SmokeTestKind)> {
+        // Find the source location of the error
+        let (_, loc_opt) = self.get_locations(error.position);
+        let source_loc_opt = self.to_proper_source_location(loc_opt).and_then(|loc| {
+            if error.kind.is_from_verification() {
+                Some(loc)
+            } else {
+                None
+            }
+        });
+        source_loc_opt.as_ref()?;
+
+        // Find the function the error was raised in
+        let fun_env_opt = self.env.get_enclosing_function(source_loc_opt.unwrap());
+        fun_env_opt.as_ref()?;
+
+        // Return whether the function of the error was in smoke test mode
+        let fun_env = fun_env_opt.unwrap();
+        if fun_env.is_pragma_true(SMOKE_TEST_PRAGMA, || false) {
+            let fname_symbol = fun_env.get_name();
+            // Determine the type smoke test this error is for
+            // Currently the only type of test is the aborts_if true check
+            let kind = SmokeTestKind::AbortsIfTrue;
+            Some((fname_symbol, kind))
+        } else {
+            None
+        }
+    }
+
+    /// Helper function for adding smoke test errors as Diagnostics.
+    fn add_smoke_test_errors(&self, funs_with_errors: HashSet<(Symbol, SmokeTestKind)>) {
```

Comment:
I wonder whether we would be matching every function with a simple name in the set of modules with this approach, even if only one of those has the pragma set.

This deserves at least some comment. It's probably ok since most of the time we have only one module here. On the other hand, it is possible to inject precise FunId and ModuleId into models to make this exact, like someone would expect using a pragma on a specific function. 

Please add a TODO and comment that this should be revisited, or correct if I'm wrong that the error identification is ambigious.

Reply:
I agree that we should inject precise FunIds and ModuleIds into the models (that's exactly what I was trying to look for originally). I've placed a comment for now.